pkcs7/smime.c

```diff
@@ -0,0 +1,249 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA
+ */
+
+#include <wolfssl/options.h>
+#include <wolfssl/wolfcrypt/settings.h>
+#include <wolfssl/ssl.h>
+#include <wolfssl/wolfcrypt/pkcs7.h>
+#include <wolfssl/wolfcrypt/error-crypt.h>
+#include <wolfssl/wolfcrypt/logging.h>
+
+
+#ifdef HAVE_SMIME
+
+/* Create an SMIME bundle, uses steps similar to PKCS7_sign() but this is
+ * with native PKCS7 API allowing for control over the devID set */
+static int Create(byte* smime, int* smimeSz, byte* key, int keySz,
+    byte* signer, int signerSz, byte* content, int contentSz, int detatched)
+{
+    WOLFSSL_BIO *out, *in;
+    WOLFSSL_PKCS7* pkcs7Compat;
+    PKCS7* pkcs7;
+    int ret = 0;
+
+    out = wolfSSL_BIO_new(wolfSSL_BIO_s_mem());
+    in  = wolfSSL_BIO_new(wolfSSL_BIO_s_mem());
+    if (out == NULL || in == NULL) {
+        printf("Failed to create bio's\n");
+        return -1;
+    }
+
+    pkcs7Compat = (WOLFSSL_PKCS7*)wolfSSL_PKCS7_new();
+    if (pkcs7Compat == NULL) {
+        ret = MEMORY_E;
+    }
+    else {
+        pkcs7 = &(pkcs7Compat->pkcs7);
+
+        /* can change devID here and set signer */
+        ret = wc_PKCS7_Init(pkcs7, NULL, INVALID_DEVID);
+    }
+
+    if (ret == 0) {
+        ret = wc_PKCS7_InitWithCert(pkcs7, signer, signerSz);
+    }
+
+    if (ret == 0) {
+        /* set signer private key, data types, defaults */
+        pkcs7->privateKey   = key;
+        pkcs7->privateKeySz = keySz;
+        pkcs7->contentOID   = DATA;  /* inner content default is DATA */
+        pkcs7->hashOID      = SHA256h;  /* default to SHA-256 hash type */
+
+        /* type of SMIME */
+        pkcs7Compat->type = SIGNED_DATA;
+
+        /* add additional chain certs if provided */
+        /* wc_PKCS7_AddCertificate(pkcs7, additionalCert, additionalCertSz); */
+
+        /* set detached flag */
+        if (detatched & PKCS7_DETACHED) {
+            ret = wc_PKCS7_SetDetached(pkcs7, 1);
+        }
+
+        /* setup content to sign */
+        if (ret == 0) {
+            if (wolfSSL_BIO_write(in, content, contentSz) != contentSz) {
+                ret = -1;
+            }
+        }
+
+        if (ret == 0) {
+            if (wolfSSL_SMIME_write_PKCS7(out, (PKCS7*)pkcs7Compat, in,
+                    detatched) != WOLFSSL_SUCCESS) {
+                printf("SMIME write failed!\n");
+                ret = -1;
+            }
+        }
+    }
+
+    if (ret == 0) {
+        ret = wolfSSL_BIO_read(out, smime, *smimeSz);
+        if (ret > 0) {
+            if (ret == *smimeSz) {
+                printf("output smime buffer too small\n");
```

Comment:
I'm getting this error when I run this example using the client key and cert files from the certs directory:

```
$ cd wolfssl-examples/pkcs7
$ make
$ ./smime ../certs/client-key.der ../certs/client-cert.der
...
output to file ./smime-created.p7s

output smime buffer too small
```

Maybe we need to increase a default buffer size?  Or detect a size somewhere?